org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/preferences/internal/DefaultModelPersistence.java

```diff
@@ -20,120 +21,149 @@
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Properties;
-import java.util.Set;
 
 import org.gradle.internal.UncheckedException;
 
 import com.google.common.base.Charsets;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
 import com.google.common.collect.Maps;
 import com.google.common.io.Files;
 
 import org.eclipse.core.resources.IProject;
 
 import org.eclipse.buildship.core.CorePlugin;
+import org.eclipse.buildship.core.GradlePluginsRuntimeException;
 import org.eclipse.buildship.core.event.Event;
 import org.eclipse.buildship.core.event.EventListener;
 import org.eclipse.buildship.core.preferences.ModelPersistence;
 import org.eclipse.buildship.core.preferences.PersistentModel;
 import org.eclipse.buildship.core.workspace.ProjectDeletedEvent;
 import org.eclipse.buildship.core.workspace.ProjectMovedEvent;
+import org.eclipse.buildship.core.workspace.WorkbenchShutdownEvent;
 
 /**
  * Default implementation for {@link PersistentModel}.
  *
  * @author Donat Csikos
  */
-public final class DefaultModelPersistence implements ModelPersistence, EventListener  {
+public final class DefaultModelPersistence implements ModelPersistence, EventListener {
+
+    // Cache holding (project name - persistent model property) pairs.
+    private final LoadingCache<String, Map<String, String>> modelCache;
```

Comment:
I think we could improve perceived performance slightly by launching a background job at startup which loads all preferences into memory.